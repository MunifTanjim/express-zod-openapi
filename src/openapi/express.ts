import type { Application, Router } from 'express'
import type { RegisteredExpressOpenAPIPlugin } from '../plugins'
import type { HttpMethod } from '../types'
import type { OpenAPISpecification } from './specification'

const enum StackName {
  QUERY = 'query',
  EXPRESS_INIT = 'expressInit',
  BOUND_DISPATCH = 'bound dispatch',
  ROUTER = 'router',
  MOUNTED_APP = 'mounted_app',
}

interface Key {
  name: string | number
  prefix: string
  suffix: string
  pattern: string
  modifier: string
}

interface Layer {
  handle: Router | ((...params: any[]) => any)
  params?: Record<string, any>
  path?: string
  keys: Key[]
  regexp: RegExp & { fast_star?: boolean; fast_slash?: boolean }
}

interface StackLayer extends Layer {
  name: StackName
  route?: Route
}

interface RouteLayer extends Layer {
  name: string
  method: HttpMethod
}

export interface Route {
  path: string | string[]
  stack: RouteLayer[]
  methods: {
    [method in HttpMethod | '_all']?: boolean
  }
}

const regexPattern = {
  param: /\(\?:\(\[\^\\\/]\+\?\)\)/,
  path: /^\^\\\/(?:(:?[\w\\.-]*(?:\\\/:?[\w\\.-]*)*)|(\(\?:\(\[\^\\\/]\+\?\)\)))\\/,
  rootPath: /^\^\\\/\?\(\?=\\\/\|\$\)$/,
}

const getPathString = (
  pathRegex: RegExp,
  keys: Layer['keys'],
): string | null => {
  const pathRegexpSource = pathRegex.source

  let regexString = pathRegexpSource

  let regexStringExecArray = regexPattern.path.exec(regexString)

  let paramIndex = 0
  while (regexPattern.param.test(regexString)) {
    const param = `:${keys[paramIndex].name}`

    regexString = regexString.replace(regexPattern.param, param)

    paramIndex++
  }

  if (regexString !== pathRegexpSource) {
    regexStringExecArray = regexPattern.path.exec(regexString)
  }

  const pathString = regexStringExecArray?.[1].replace(/\\\//g, '/')

  return pathString ?? null
}

const processRoute = (
  spec: OpenAPISpecification,
  route: Route,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[],
): void => {
  const routePaths = Array.isArray(route.path) ? route.path : [route.path]

  routePaths.forEach((routePath) => {
    const path = `${
      basePath ? `${basePath}${routePath === '/' ? '' : routePath}` : routePath
    }`.replace(/:([a-zA-Z0-9_]+)/g, '{$1}')

    const methods = Object.keys(route.methods).filter(
      (method): method is HttpMethod => method !== '_all',
    )

    if (!spec.paths[path]) {
      spec.setPathItem(path, {})
    }

    for (const method of methods) {
      spec.setPathItemOperation(path, method, {
        responses: {
          default: {
            description: '',
          },
        },
      })

      for (const plugin of plugins) {
        const stashedValue = plugin.stash.find(route, method)
        if (stashedValue) {
          plugin.processRoute(plugin.specification, stashedValue, {
            path,
            method,
          })
        }
      }
    }
  })
}

export const processExpressRouters = (
  spec: OpenAPISpecification,
  app: Application | Router,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[],
): void => {
  const stack: StackLayer[] = app.stack ?? (app as Application)._router?.stack

  if (!stack) {
    spec.setPathItem(basePath, {})

    return
  }

  for (const layer of stack) {
    // terminal route
    if (layer.route) {
      processRoute(spec, layer.route, basePath, plugins)

      continue
    }

    if (
      layer.name !== StackName.ROUTER &&
      layer.name !== StackName.BOUND_DISPATCH &&
      layer.name !== StackName.MOUNTED_APP
    ) {
      continue
    }

    if (regexPattern.path.test(layer.regexp.source)) {
      const parsedPath = getPathString(layer.regexp, layer.keys)

      if (parsedPath) {
        processExpressRouters(
          spec,
          layer.handle as Router,
          basePath + '/' + parsedPath,
          plugins,
        )
      }

      continue
    }

    if (
      !layer.path &&
      layer.regexp &&
      !regexPattern.rootPath.test(layer.regexp.source)
    ) {
      const regexPath = ' RegExp(' + layer.regexp + ') '

      processExpressRouters(
        spec,
        layer.handle as Router,
        basePath + '/' + regexPath,
        plugins,
      )

      continue
    }

    processExpressRouters(spec, layer.handle as Router, basePath, plugins)
  }
}
