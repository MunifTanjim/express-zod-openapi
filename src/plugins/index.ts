import type { Handler } from 'express'
import type { OpenAPISpecification } from '../openapi'
import type { Stash } from '../stash'
import type { HttpMethod } from '../types'

export type ExpressOpenAPIPluginInternals<StashValue> = {
  specification: OpenAPISpecification
  stash: Stash<StashValue>
}

export type RouteProcessor<StashValue> = (
  specification: OpenAPISpecification,
  stash: StashValue,
  info: {
    path: string
    method: HttpMethod
  },
) => void

export interface ExpressOpenAPIPlugin<
  StashValue = unknown,
  GetMiddleware extends (...params: any[]) => Handler | Handler[] = (
    ...params: any[]
  ) => Handler,
> {
  name: string
  getMiddleware: (
    internals: ExpressOpenAPIPluginInternals<StashValue>,
    ...params: Parameters<GetMiddleware>
  ) => ReturnType<GetMiddleware>
  processRoute: RouteProcessor<StashValue>
}

export interface RegisteredExpressOpenAPIPlugin {
  name: string
  specification: OpenAPISpecification
  stash: Stash<any>
  processRoute: RouteProcessor<any>
}
