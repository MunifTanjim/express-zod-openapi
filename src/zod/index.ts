export { zodToJsonSchema } from 'zod-to-json-schema'
export { RequestValidationError, ResponseValidationError } from './error'
