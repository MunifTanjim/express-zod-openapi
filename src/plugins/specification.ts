import {
  Handler,
  Request as ExpressRequest,
  RequestHandler as ExpressRequestHandler,
  Response as ExpressResponse,
} from 'express'
import {
  HeaderObject,
  HeadersObject,
  ParameterLocation,
  ParameterObject,
  RequestBodyObject,
  ResponseObject,
} from 'openapi3-ts'
import { Schema, TypeOf, ZodError } from 'zod'
import { zodToJsonSchema } from '../zod'
import { ExpressOpenAPIPlugin } from './types'

export type RequestSegment =
  | 'body'
  | 'cookies'
  | 'headers'
  | 'params'
  | 'query'
  | 'signedCookies'

export type ResponseSegment = 'body' | 'headers'

type SpecInfo = {
  operationId?: string
  req: Map<RequestSegment, Schema<any> | undefined>
  res: Map<
    string,
    | {
        [segment in ResponseSegment]?: Schema<any>
      }
    | undefined
  >
}

type SpecificationSchema = {
  operationId?: string
  req?: {
    [segment in RequestSegment]?: Schema<any>
  }
  res?: {
    [statusCode: string]: {
      [segment in ResponseSegment]?: Schema<any>
    }
  }
}

type UnionTypeOf<T extends Schema<any>> = T extends unknown ? TypeOf<T> : never

type ResponseSegmentType<
  VS extends SpecificationSchema,
  S extends ResponseSegment,
  D = unknown
> = VS['res'] extends undefined
  ? D
  : NonNullable<VS['res']>[keyof NonNullable<VS['res']>][S] extends undefined
  ? D
  : UnionTypeOf<
      NonNullable<NonNullable<VS['res']>[keyof NonNullable<VS['res']>][S]>
    >

type RequestSegmentType<
  VS extends SpecificationSchema,
  S extends RequestSegment,
  D = unknown
> = VS['req'] extends undefined
  ? D
  : NonNullable<VS['req']>[S] extends undefined
  ? D
  : TypeOf<NonNullable<NonNullable<VS['req']>[S]>>

export type Response<S extends SpecificationSchema> = ExpressResponse<
  ResponseSegmentType<S, 'body'>
>

export type Request<S extends SpecificationSchema> = ExpressRequest<
  RequestSegmentType<S, 'params'>,
  ResponseSegmentType<S, 'body'>,
  RequestSegmentType<S, 'body'>,
  RequestSegmentType<S, 'query'>
>

export type RequestHandler<S extends SpecificationSchema> =
  ExpressRequestHandler<
    RequestSegmentType<S, 'params'>,
    ResponseSegmentType<S, 'body'>,
    RequestSegmentType<S, 'body'>,
    RequestSegmentType<S, 'query'>
  >

export type GetSpecificationMiddleware = (
  schema: SpecificationSchema
) => Handler

type SpecificationPlugin = ExpressOpenAPIPlugin<
  SpecInfo,
  GetSpecificationMiddleware
>

export class RequestValidationError extends Error {
  segment: RequestSegment
  validationError: ZodError

  constructor(validationError: ZodError, segment: RequestSegment) {
    super(validationError.message)

    this.segment = segment
    this.validationError = validationError
  }
}

export class ResponseValidationError extends Error {
  segment: ResponseSegment
  validationError: ZodError

  constructor(validationError: ZodError, segment: ResponseSegment) {
    super(validationError.message)

    this.segment = segment
    this.validationError = validationError
  }
}

const parameterLocationBySegment: {
  [key in Exclude<RequestSegment, 'body'>]: ParameterLocation
} = {
  cookies: 'cookie',
  signedCookies: 'cookie',
  headers: 'header',
  params: 'path',
  query: 'query',
}

async function validateRequest(
  req: ExpressRequest,
  specInfo: SpecInfo,
  segmentOrder: RequestSegment[]
) {
  await segmentOrder.reduce((promise, segment) => {
    return promise.then(async () => {
      const schema = specInfo.req.get(segment)

      if (!schema) {
        return null
      }

      const result = await schema.safeParseAsync(req[segment])

      if (!result.success) {
        throw new RequestValidationError(result.error, segment)
      }

      req[segment] = result.data

      return null
    })
  }, Promise.resolve(null))
}

function setupResponseValidation(
  res: ExpressResponse,
  specInfo: SpecInfo,
  segmentOrder: ResponseSegment[],
  next: (err?: unknown) => void
) {
  const originalSend = res.send

  res.send = function validateAndSendResponse(...args) {
    res.send = originalSend

    const body = args[0]

    const isJsonContent = /application\/json/.test(
      String(res.get('content-type'))
    )

    const value: { [key in ResponseSegment]: unknown } = {
      body: isJsonContent ? JSON.parse(body) : body,
      headers: res.getHeaders(),
    }

    const schemaBySegment = specInfo.res.has(String(res.statusCode))
      ? specInfo.res.get(String(res.statusCode))
      : specInfo.res.get('default')

    if (!schemaBySegment) {
      next(
        new Error(`Validation Schema not found for Response(${res.statusCode})`)
      )
      return res
    }

    for (const segment of segmentOrder) {
      const schema = schemaBySegment[segment]

      if (!schema) {
        continue
      }

      const result = schema.safeParse(value[segment])

      if (!result.success) {
        next(new ResponseValidationError(result.error, segment))
        return res
      }

      if (segment === 'body') {
        value[segment] = result.data
      }
    }

    return originalSend.apply(res, [
      isJsonContent ? JSON.stringify(value.body) : value.body,
    ])
  }
}

export const getSpecificationPlugin = ({
  req: reqConfig = {},
  res: resConfig = {},
}: {
  req?: {
    segmentOrder?: RequestSegment[]
    skipValidation?: boolean
  }
  res?: {
    segmentOrder?: ResponseSegment[]
    skipValidation?: boolean
  }
} = {}): SpecificationPlugin => {
  const {
    segmentOrder: reqSegmentOrder = [
      'headers',
      'params',
      'query',
      'cookies',
      'signedCookies',
      'body',
    ],
    skipValidation: skipRequestValidation = false,
  } = reqConfig

  const {
    segmentOrder: resSegmentOrder = ['body', 'headers'],
    skipValidation: skipResponseValidation = false,
  } = resConfig

  const specificationPlugin: SpecificationPlugin = {
    name: 'zod-openapi-spec',

    getMiddleware: (internals, validationSchema): Handler => {
      const specInfo: SpecInfo = {
        operationId: validationSchema.operationId,
        req: new Map(),
        res: new Map(),
      }

      for (const segment of reqSegmentOrder) {
        const schema = validationSchema.req?.[segment]
        if (schema) {
          specInfo.req.set(segment, schema)
        }
      }

      for (const [code, schemaBySegment] of Object.entries(
        validationSchema.res ?? {}
      )) {
        specInfo.res.set(
          code,
          Object.entries(schemaBySegment).reduce<{
            [key in ResponseSegment]?: Schema<unknown>
          }>((bySegment, [segment, schema]) => {
            if (schema) {
              bySegment[segment as ResponseSegment] = schema
            }
            return bySegment
          }, {})
        )
      }

      const validationMiddleware: Handler = async (
        req,
        res,
        next
      ): Promise<void> => {
        if (!skipResponseValidation && validationSchema.res) {
          setupResponseValidation(res, specInfo, resSegmentOrder, next)
        }

        if (!skipRequestValidation && validationSchema.req) {
          try {
            await validateRequest(req, specInfo, reqSegmentOrder)
          } catch (err) {
            next(err)
            return
          }
        }

        next()
      }

      internals.stash.store(validationMiddleware, specInfo)

      return validationMiddleware
    },

    processRoute: (specification, specInfo, { path, method }): void => {
      for (const [segment, schema] of specInfo.req.entries()) {
        if (!schema) {
          continue
        }

        if (specInfo.operationId) {
          specification.setPathItemOperation(path, method, {
            ...specification.getPathItemOperation(path, method)!,
            operationId: specInfo.operationId,
          })
        }

        if (segment === 'body') {
          const result = zodToJsonSchema(schema)

          const requestBody: RequestBodyObject = {
            content: {
              'application/json': {
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                schema: result,
              },
            },
          }

          specification.setPathItemOperationRequestBody(
            path,
            method,
            requestBody
          )

          continue
        }

        const location = parameterLocationBySegment[segment]

        const result = zodToJsonSchema(schema)

        if ('properties' in result) {
          const { properties, required = [] } = result

          if (properties) {
            for (const [name, schema] of Object.entries(properties)) {
              const parameterObject: ParameterObject = {
                name,
                in: location,
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-expect-error
                schema: schema,
                required: required.includes(name),
              }

              specification.addPathItemOperationParameter(
                path,
                method,
                parameterObject
              )
            }
          }
        }
      }

      for (const [key, schemaBySegment] of specInfo.res.entries()) {
        if (!schemaBySegment) {
          continue
        }

        const httpStatusCode = key as number | 'default'

        if (schemaBySegment.body) {
          const result = zodToJsonSchema(schemaBySegment.body)

          const responseObject: ResponseObject = {
            ...specification.getPathItemOperationResponse(
              path,
              method,
              httpStatusCode
            ),
            description: '',
            content: {
              'application/json': {
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                schema: result,
              },
            },
          }

          specification.setPathItemOperationResponse(
            path,
            method,
            httpStatusCode,
            responseObject
          )
        }

        if (schemaBySegment.headers) {
          const result = zodToJsonSchema(schemaBySegment.headers)

          if ('properties' in result) {
            const { properties, required = [] } = result

            const headersObject: HeadersObject = {}

            if (properties) {
              for (const [name, schema] of Object.entries(properties)) {
                const headerObject: HeaderObject = {
                  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                  // @ts-ignore
                  schema: schema,
                  required: required.includes(name),
                }

                headersObject[name] = headerObject
              }
            }

            const responseObject: ResponseObject = {
              ...specification.getPathItemOperationResponse(
                path,
                method,
                httpStatusCode
              ),
              description: '',
              headers: headersObject,
            }

            specification.setPathItemOperationResponse(
              path,
              method,
              httpStatusCode,
              responseObject
            )
          }
        }
      }
    },
  }

  return specificationPlugin
}
