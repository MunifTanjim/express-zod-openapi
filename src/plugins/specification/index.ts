import type { Handler } from 'express'
import type {
  HeaderObject,
  HeadersObject,
  ParameterLocation,
  ParameterObject,
  RequestBodyObject,
  ResponseObject,
} from 'openapi3-ts'
import type { Schema } from 'zod'
import type { ExpressOpenAPIPlugin, RouteProcessor } from '..'
import type { RequestSegment, ResponseSegment } from '../../types'
import { setupResponseValidation, validateRequest } from '../../utils'
import { zodToJsonSchema } from '../../zod'
import type { SpecificationSchema } from './types'

export type {
  Request,
  RequestHandler,
  Response,
  SpecificationSchema,
} from './types'

type SpecInfo = {
  operationId?: string
  req: Map<RequestSegment, Schema<any> | undefined>
  res: Map<
    string,
    | {
        [segment in ResponseSegment]?: Schema<any>
      }
    | undefined
  >
}

export type GetSpecificationMiddleware = (
  schema: SpecificationSchema,
) => Handler

type SpecificationPlugin = ExpressOpenAPIPlugin<
  SpecInfo,
  GetSpecificationMiddleware
>

const parameterLocationBySegment: {
  [key in Exclude<RequestSegment, 'body'>]: ParameterLocation
} = {
  cookies: 'cookie',
  signedCookies: 'cookie',
  headers: 'header',
  params: 'path',
  query: 'query',
}

const routeProcessor: RouteProcessor<SpecInfo> = (
  specification,
  specInfo,
  { path, method },
) => {
  for (const [segment, schema] of specInfo.req.entries()) {
    if (!schema) {
      continue
    }

    if (specInfo.operationId) {
      specification.setPathItemOperation(path, method, {
        ...specification.getPathItemOperation(path, method)!,
        operationId: specInfo.operationId,
      })
    }

    if (segment === 'body') {
      const result = zodToJsonSchema(schema)

      const requestBody: RequestBodyObject = {
        content: {
          'application/json': {
            // @ts-expect-error ...
            schema: result,
          },
        },
      }

      specification.setPathItemOperationRequestBody(path, method, requestBody)

      continue
    }

    const location = parameterLocationBySegment[segment]

    const result = zodToJsonSchema(schema)

    if ('properties' in result) {
      const { properties, required = [] } = result

      if (properties) {
        for (const [name, schema] of Object.entries(properties)) {
          const parameterObject: ParameterObject = {
            name,
            in: location,
            // @ts-expect-error ...
            schema: schema,
            required: required.includes(name),
          }

          specification.addPathItemOperationParameter(
            path,
            method,
            parameterObject,
          )
        }
      }
    }
  }

  for (const [key, schemaBySegment] of specInfo.res.entries()) {
    if (!schemaBySegment) {
      continue
    }

    const httpStatusCode = key as number | 'default'

    if (schemaBySegment.body) {
      const result = zodToJsonSchema(schemaBySegment.body)

      const responseObject: ResponseObject = {
        ...specification.getPathItemOperationResponse(
          path,
          method,
          httpStatusCode,
        ),
        description: '',
        content: {
          'application/json': {
            // @ts-expect-error ...
            schema: result,
          },
        },
      }

      specification.setPathItemOperationResponse(
        path,
        method,
        httpStatusCode,
        responseObject,
      )
    }

    if (schemaBySegment.headers) {
      const result = zodToJsonSchema(schemaBySegment.headers)

      if ('properties' in result) {
        const { properties, required = [] } = result

        const headersObject: HeadersObject = {}

        if (properties) {
          for (const [name, schema] of Object.entries(properties)) {
            const headerObject: HeaderObject = {
              // @ts-expect-error ...
              schema: schema,
              required: required.includes(name),
            }

            headersObject[name] = headerObject
          }
        }

        const responseObject: ResponseObject = {
          ...specification.getPathItemOperationResponse(
            path,
            method,
            httpStatusCode,
          ),
          description: '',
          headers: headersObject,
        }

        specification.setPathItemOperationResponse(
          path,
          method,
          httpStatusCode,
          responseObject,
        )
      }
    }
  }
}

const create = ({
  req: reqConfig = {},
  res: resConfig = {},
}: {
  req?: {
    segmentOrder?: RequestSegment[]
    skipValidation?: boolean
  }
  res?: {
    segmentOrder?: ResponseSegment[]
    skipValidation?: boolean
  }
} = {}): SpecificationPlugin => {
  const {
    segmentOrder: reqSegmentOrder = [
      'headers',
      'params',
      'query',
      'cookies',
      'signedCookies',
      'body',
    ],
    skipValidation: skipRequestValidation = false,
  } = reqConfig

  const {
    segmentOrder: resSegmentOrder = ['body', 'headers'],
    skipValidation: skipResponseValidation = false,
  } = resConfig

  const specificationPlugin: SpecificationPlugin = {
    name: 'zod-openapi-spec',

    getMiddleware: (internals, validationSchema): Handler => {
      const specInfo: SpecInfo = {
        operationId: validationSchema.operationId,
        req: new Map(),
        res: new Map(),
      }

      for (const segment of reqSegmentOrder) {
        const schema = validationSchema.req?.[segment]
        if (schema) {
          specInfo.req.set(segment, schema)
        }
      }

      for (const [code, schemaBySegment] of Object.entries(
        validationSchema.res ?? {},
      )) {
        specInfo.res.set(
          code,
          Object.entries(schemaBySegment).reduce<{
            [key in ResponseSegment]?: Schema<unknown>
          }>((bySegment, [segment, schema]) => {
            if (schema) {
              bySegment[segment as ResponseSegment] = schema
            }
            return bySegment
          }, {}),
        )
      }

      const validationMiddleware: Handler = async (
        req,
        res,
        next,
      ): Promise<void> => {
        if (!skipResponseValidation && validationSchema.res) {
          setupResponseValidation(res, specInfo.res, resSegmentOrder, next)
        }

        if (!skipRequestValidation && validationSchema.req) {
          try {
            await validateRequest(req, specInfo.req, reqSegmentOrder)
          } catch (err) {
            next(err)
            return
          }
        }

        next()
      }

      internals.stash.store(validationMiddleware, specInfo)

      return validationMiddleware
    },

    processRoute: routeProcessor,
  }

  return specificationPlugin
}

/**
 * @deprecated Use `SpecificationPlugin.create`
 */
export const getSpecificationPlugin = create

export const SpecificationPlugin = {
  create,
  routeProcessor,
}
