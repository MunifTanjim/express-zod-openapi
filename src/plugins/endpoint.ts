import {
  Router,
  type Request as ExpressRequest,
  type Response as ExpressResponse,
  type Handler,
  type NextFunction,
} from 'express'
import { z, type Schema, type TypeOf } from 'zod'
import type { HttpMethod } from '../types'
import { setupResponseValidation, validateRequest } from '../utils'
import { SpecificationPlugin } from './specification'
import type {
  ExpressOpenAPIPlugin,
  RequestSegment,
  ResponseSegment,
  RouteProcessor,
} from './types'

type SpecInfo = {
  defaultStatusCode: number
  operationId?: string
  req: Map<RequestSegment, Schema<any> | undefined>
  res: Map<
    string,
    | {
        [segment in ResponseSegment]?: Schema<any>
      }
    | undefined
  >
}

type SpecificationSchema = {
  operationId?: string
  req?: {
    [segment in RequestSegment]?: Schema<any>
  }
  res?: {
    [statusCode: string]: {
      [segment in ResponseSegment]?: Schema<any>
    }
  }
}

type UnionTypeOf<T extends Schema<any>> = T extends unknown ? TypeOf<T> : never

type ResponseSegmentType<
  VS extends SpecificationSchema,
  S extends ResponseSegment,
  D = unknown,
> = VS['res'] extends undefined
  ? D
  : NonNullable<VS['res']>[keyof NonNullable<VS['res']>][S] extends undefined
  ? D
  : UnionTypeOf<
      NonNullable<NonNullable<VS['res']>[keyof NonNullable<VS['res']>][S]>
    >

type RequestSegmentType<
  VS extends SpecificationSchema,
  S extends RequestSegment,
  D = unknown,
> = VS['req'] extends undefined
  ? D
  : NonNullable<VS['req']>[S] extends undefined
  ? D
  : TypeOf<NonNullable<NonNullable<VS['req']>[S]>>

type Response<S extends SpecificationSchema> = ExpressResponse<
  ResponseSegmentType<S, 'body'>
>

type Request<S extends SpecificationSchema> = ExpressRequest<
  RequestSegmentType<S, 'params'>,
  ResponseSegmentType<S, 'body'>,
  RequestSegmentType<S, 'body'>,
  RequestSegmentType<S, 'query'>
>

type MaybePromise<T> = Promise<T> | T

const specInfoSymbol = Symbol('zod-openapi-endpoint-specInfo')

type SpecificationChain<
  Schema extends SpecificationSchema,
  Called extends '' | 'req' | 'res',
> = Handler[] &
  Omit<
    {
      req: <T extends NonNullable<SpecificationSchema['req']>>(
        schema: T,
      ) => SpecificationChain<Schema & { req: T }, Called | 'req'>
      res: <T extends NonNullable<SpecificationSchema['res']>>(
        schema: T,
      ) => SpecificationChain<Schema & { res: T }, Called | 'res'>
    },
    Called
  > & {
    [specInfoSymbol]: SpecInfo
    router: Router
    handle: (
      handler: (
        req: Request<Schema>,
        res: Response<Schema>,
        next: NextFunction,
      ) =>
        | MaybePromise<ResponseSegmentType<Schema, 'body'>>
        | MaybePromise<void>,
    ) => SpecificationChain<Schema, Called | 'req' | 'res'>
  }

function SpecChainReq<
  S extends SpecificationSchema,
  C extends '' | 'req' | 'res',
  T extends NonNullable<SpecificationSchema['req']>,
>(this: SpecificationChain<S, C>, spec: T) {
  const specInfo = this[specInfoSymbol]

  for (const [segment, schema] of Object.entries(spec)) {
    specInfo.req.set(segment as RequestSegment, schema)
  }

  return this as unknown as SpecificationChain<S & { req: T }, C | 'req'>
}

function SpecChainRes<
  S extends SpecificationSchema,
  C extends '' | 'req' | 'res',
  T extends NonNullable<SpecificationSchema['res']>,
>(this: SpecificationChain<S, C>, spec: T) {
  const specInfo = this[specInfoSymbol]

  for (const [code, schemaBySegment] of Object.entries(spec)) {
    const statusCode = Number(code)
    if (200 < statusCode && statusCode < 300) {
      specInfo.defaultStatusCode = statusCode
    }

    specInfo.res.set(
      code,
      Object.entries(schemaBySegment).reduce<{
        [key in ResponseSegment]?: Schema<unknown>
      }>((bySegment, [segment, schema]) => {
        if (schema) {
          bySegment[segment as ResponseSegment] = schema
        }
        return bySegment
      }, {}),
    )
  }

  if (specInfo.res.size > 0 && !specInfo.res.has('default')) {
    specInfo.res.set('default', { body: z.any() })
  }

  return this as unknown as SpecificationChain<S & { res: T }, C | 'res'>
}

function SpecChainHandle<
  S extends SpecificationSchema,
  C extends '' | 'req' | 'res',
>(
  this: SpecificationChain<S, C>,
  handler: (
    req: Request<S>,
    res: Response<S>,
    next: NextFunction,
  ) => MaybePromise<ResponseSegmentType<S, 'body'>> | MaybePromise<void>,
) {
  if (handler.length === 3) {
    this.router.use(handler)
  } else {
    this.router.use(async (req, res, next) => {
      const response = await handler(req, res, next)
      if (res.headersSent) {
        return
      }

      if (typeof response === 'undefined') {
        return next()
      }

      if (res.statusCode === 200) {
        res.statusCode = this[specInfoSymbol].defaultStatusCode
      }

      res.json(response)
    })
  }

  return this as SpecificationChain<S, C | 'req' | 'res'>
}

function SpecChain(
  getRouter: (specInfo: SpecInfo) => Router,
): SpecificationChain<Record<string, never>, ''> {
  const specInfo: SpecInfo = {
    defaultStatusCode: 200,
    req: new Map(),
    res: new Map(),
  }

  const chain = Object.assign([], {
    [specInfoSymbol]: specInfo,
    router: getRouter(specInfo),
    req: SpecChainReq,
    res: SpecChainRes,
    handle: SpecChainHandle,
  })

  return chain
}

export type GetEndpointMiddleware = (
  method: HttpMethod,
  path: string,
) => ReturnType<typeof SpecChain>

type EndpointPlugin = ExpressOpenAPIPlugin<SpecInfo, GetEndpointMiddleware>

const routeProcessor: RouteProcessor<SpecInfo> = (
  specification,
  specInfo,
  { method, path },
) => {
  SpecificationPlugin.routeProcessor(specification, specInfo, {
    method,
    path,
  })
}

const create = ({
  req: reqConfig = {},
  res: resConfig = {},
}: {
  req?: {
    segmentOrder?: RequestSegment[]
    skipValidation?: boolean
  }
  res?: {
    segmentOrder?: ResponseSegment[]
    skipValidation?: boolean
  }
} = {}) => {
  const {
    segmentOrder: reqSegmentOrder = [
      'headers',
      'params',
      'query',
      'cookies',
      'signedCookies',
      'body',
    ],
    skipValidation: skipRequestValidation = false,
  } = reqConfig

  const {
    segmentOrder: resSegmentOrder = ['body', 'headers'],
    skipValidation: skipResponseValidation = false,
  } = resConfig

  const router = Router()

  const endpointPlugin: EndpointPlugin = {
    name: 'zod-openapi-endpoint-plugin',

    getMiddleware: (internals, method, path) => {
      return SpecChain((specInfo) => {
        const validationMiddleware: Handler = async (
          req,
          res,
          next,
        ): Promise<void> => {
          if (!skipResponseValidation && specInfo.res.size) {
            setupResponseValidation(res, specInfo.res, resSegmentOrder, next)
          }

          if (!skipRequestValidation && specInfo.req.size) {
            try {
              await validateRequest(req, specInfo.req, reqSegmentOrder)
            } catch (err) {
              next(err)
              return
            }
          }

          next()
        }

        internals.stash.store(validationMiddleware, specInfo)

        return router[method](path, validationMiddleware)
      })
    },

    processRoute: routeProcessor,
  }

  return Object.assign(endpointPlugin, { router })
}

export const EndpointPlugin = {
  create,
  routeProcessor,
}
