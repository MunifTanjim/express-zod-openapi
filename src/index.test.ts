import express from 'express'
import request from 'supertest'
import { z } from 'zod'
import { ExpressOpenAPI } from './index'
import { getSpecificationPlugin } from './plugins'

const preprocessors = {
  numberString: (v: unknown) => {
    const val = Number(v)
    return Number.isNaN(val) ? v : val
  },
}

describe('ExpressOpenAPI', () => {
  describe('constructor', () => {
    test('can create instance without arguments', () => {
      expect(new ExpressOpenAPI()).toBeInstanceOf(ExpressOpenAPI)
    })

    test('can create instance with arguments', () => {
      expect(new ExpressOpenAPI({ useStringStashKey: true })).toBeInstanceOf(
        ExpressOpenAPI,
      )
    })
  })

  test('basic request schema generation', () => {
    const expressOpenApi = new ExpressOpenAPI()

    const getSpecificationMiddleware = expressOpenApi.registerPlugin(
      getSpecificationPlugin(),
    )

    const app = express()

    app.post(
      '/ping',
      getSpecificationMiddleware({
        operationId: 'ping',
        req: {
          query: z.object({
            count: z.preprocess(
              preprocessors.numberString,
              z
                .number({ description: 'asd' })
                .int()
                .describe('number of times ping will pong'),
            ),
          }),
          body: z.object({
            type: z.enum(['plastic', 'rubber', 'wood']).default('rubber'),
          }),
        },
      }),
    )

    const specification = expressOpenApi.populateSpecification(app)

    expect(specification.paths).toMatchSnapshot()
  })

  test('basic response schema generation', async () => {
    const expressOpenApi = new ExpressOpenAPI()

    const getSpecificationMiddleware = expressOpenApi.registerPlugin(
      getSpecificationPlugin(),
    )

    const app = express()

    app.post(
      '/ping',
      getSpecificationMiddleware({
        operationId: 'ping',
        res: {
          200: {
            headers: z.object({
              'x-count': z.preprocess(
                preprocessors.numberString,
                z.number().int(),
              ),
            }),
            body: z.object({
              pong: z.number().int(),
            }),
          },
        },
      }),
      (_req, res) => {
        res.set('x-count', '42')
        res.status(200).json({ pong: 42 })
      },
    )

    const specification = expressOpenApi.populateSpecification(app)

    expect(specification.paths).toMatchSnapshot()

    const response = await request(app).post('/ping')

    expect(response.status).toBe(200)
    expect(response.body).toMatchObject({ pong: 42 })
  })
})
