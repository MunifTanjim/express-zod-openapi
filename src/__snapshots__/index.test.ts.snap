// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpressOpenAPI basic request schema generation 1`] = `
Object {
  "/ping": Object {
    "post": Object {
      "operationId": "ping",
      "parameters": Array [
        Object {
          "in": "query",
          "name": "count",
          "required": true,
          "schema": Object {
            "description": "number of times ping will pong",
            "type": "integer",
          },
        },
      ],
      "requestBody": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "additionalProperties": false,
              "properties": Object {
                "type": Object {
                  "default": "rubber",
                  "enum": Array [
                    "plastic",
                    "rubber",
                    "wood",
                  ],
                  "type": "string",
                },
              },
              "type": "object",
            },
          },
        },
      },
      "responses": Object {
        "default": Object {
          "description": "",
        },
      },
    },
  },
}
`;

exports[`ExpressOpenAPI basic response schema generation 1`] = `
Object {
  "/ping": Object {
    "post": Object {
      "responses": Object {
        "200": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "additionalProperties": false,
                "properties": Object {
                  "pong": Object {
                    "type": "integer",
                  },
                },
                "required": Array [
                  "pong",
                ],
                "type": "object",
              },
            },
          },
          "description": "",
          "headers": Object {
            "x-count": Object {
              "required": true,
              "schema": Object {
                "type": "integer",
              },
            },
          },
        },
        "default": Object {
          "description": "",
        },
      },
    },
  },
}
`;
